mitodl/mitxonline
# Add financial aid migration query and tier program mapping table

#### Pre-Flight checklist

#### What are the relevant tickets?
https://github.com/mitodl/mitxonline/issues/702

#### What's this PR do?
- Adding a Tier Program mapping table to map MicroMaster TierProgram to MITxOnline FlexiblePriceTier
- Adding a migration query to bring over the MicroMaster DEDP financial aid records to MITxOnline

#### How should this be manually tested?

- Run the migrate command to create a new mapping table locally

> docker-compose run --rm web ./manage.py migrate

- Go to http://mitxonline.odl.local:8013/admin/micromasters_import/programtierid/, and manually add the following tier program mapping:

Tier Program | MicroMaster Tier Program ID | MITxOnline Flexible Price  (This is your local flexible price instance)
-- | -- | --
Tier 1 | [51](https://micromasters-rc.odl.mit.edu/admin/financialaid/tierprogram/51/change/ )| [38](http://mitxonline.odl.local:8013//admin/flexiblepricing/flexiblepricetier/38/change/)
Tier 2 | [52](https://micromasters-rc.odl.mit.edu/admin/financialaid/tierprogram/52/change/) | [39](https://mitxonline.odl.local:8013/admin/flexiblepricing/flexiblepricetier/39/change/)
Tier 3 | [53](https://micromasters-rc.odl.mit.edu/admin/financialaid/tierprogram/53/change/ )| [40](https://mitxonline.odl.local:8013/admin/flexiblepricing/flexiblepricetier/40/change/)
Tier 4 | [54](https://micromasters-rc.odl.mit.edu/admin/financialaid/tierprogram/54/change/) | [41](https://mitxonline.odl.local:8013/admin/flexiblepricing/flexiblepricetier/41/change/)

- Run the following queries on your local DB to connect to our MicroMaster RC

    (Note that this step is only needed to test locally, our MITxOnline RC already has '**micromasters**' foreign data wrapper configured by DevOps)
    (Replace host/user/password with **micromasters-rc**  credential from Heroku config)

```-- load postgres_fdw extension on MITxOnline DB
CREATE EXTENSION IF NOT EXISTS postgres_fdw;

--- create a foreigner server to connect to external MicroMaster DB
CREATE SERVER IF NOT EXISTS micromaster_bridge
	FOREIGN DATA WRAPPER postgres_fdw
    OPTIONS (host '------', dbname 'micromasters');

-- setup a user mapping to authenticate:
CREATE USER MAPPING IF NOT EXISTS FOR postgres
	SERVER micromaster_bridge
    OPTIONS (user 'micromasters', password '------');

--- create a new schema on local MITxOnline DB to import external tables (from MicroMaster) for migration
DROP SCHEMA IF EXISTS micromasters CASCADE;
CREATE SCHEMA micromasters;
IMPORT FOREIGN SCHEMA public 
FROM SERVER micromaster_bridge INTO micromasters;
```

- Run the following migration commands, and make sure no errors

```docker-compose run --rm web ./manage.py import_micromasters_data```

- Verify that DEDP financial aid records are imported from MicroMaster RC

   A couple of constraints:
>   MicroMaster users need to exist on MITxOnline (We use the 'mitxonline' social auth account on MicroMaster to match MITxOnline users)
>   Only bring in financial aid records where the program is DEDP and financialaid_tierprogram.current = true.

<img width="1260" alt="image" src="https://user-images.githubusercontent.com/3138890/183491614-554aea97-5773-4ec8-af5f-f4a33a730265.png">

>  http://mitxonline.odl.local:8013/admin/flexiblepricing/flexibleprice/, 
>  http://mitxonline.odl.local:8013/staff-dashboard/flexible_pricing


#### Where should the reviewer start?
(Optional)

#### Any background context you want to provide?
(Optional)

#### Screenshots (if appropriate)
(Optional)
